setendings.py

```diff
@@ -0,0 +1,169 @@
+    if last_child:
+        assert hasattr(last_child, 'as_string'), '''ERROR:️ node {} must have the
+            .as_string property.'''.format(last_child)
+        # Some nodes have col_offset prop not set. e.g. astroid.Arguments..
+        assert last_child.col_offset is not None, '''ERROR:️ node {} last_child 
+            has col_offset == None, which we need to set node's col_offset.
+            '''.format(last_child)
+        node.col_offset = last_child.col_offset
+        # print('\n before:', node.col_offset)
+        # print('\n node:', node)
+        # print('\n col_offset1:', last_child.col_offset)
+        node.end_col_offset = last_child.col_offset + len(last_child.as_string())
+    else:  # No children..
+        assert hasattr(node, 'as_string'), '''ERROR:️ node {} must have the 
+            .as_string property.'''.format(node)
+        # Some nodes have col_offset prop not set. e.g. astroid.Arguments..
+        if node.col_offset is None:
+            # Get col_offset from first child from generator.
+            first_child = next(node.get_children())
+            # print('\n first_child:', first_child)
+            node.col_offset = first_child.col_offset
+    
+        # print('\n node:', node)
+        # print('\n col_offset2:', node.col_offset)
+        node.end_col_offset = node.col_offset + len(node.as_string())
+
+def set_end_col_offset(node, last_child=None):
+    """Set the end_col_offset property by the as_string attribute.
+    Uses the end column location of the last line, instead of the widest line.
+    Need to do a custom postorder traversal to set end_col_offset property of 
+    child nodes first, since it is sometimes unset in the last child.
+    Postorder setting of children results in the assumption: if a node has the
+    propert set, then all of its children are set.
+    """
+    if hasattr(node, 'end_lineno') and hasattr(node, 'end_col_offset'):
+        if node.end_lineno and node.end_col_offset:
+            return  # reduces runtime since properties already set.
+
+    # Check: contradiction found. Revise NO_CHILDREN_TYPE list for correctness.
+    assert not last_child or type(node) not in NO_CHILDREN_TYPE, '''ERROR:️ {} 
+        node in NO_CHILDREN_TYPE has children ({}). Suggest: remove it from the 
+        list.Context:\n{}'''.format(node, last_child, node.as_string())
+
+    if hasattr(last_child, 'end_col_offset'):  # Set by last child
+        assert last_child.end_col_offset is not None, '''ERROR:️ last child 
+            end_col_offset should not be None if used to set others. 
+            Node {}.'''.format(last_child)
+        node.end_col_offset = last_child.end_col_offset
+    # Set by string of node or last_child. May not have child.
+    else:
+        # Recall: children set first in postorder traversal.
+        # Should this always set by string, is there another way?
+        set_end_col_offset_by_string(node, last_child)
```

Comment:
If the `last_child` has no `end_col_offset` attribute set, then this node has no business having its own `end_col_offset` set.

You can think of what you're doing as building up correct `end_col_offset` from the leaves upwards in an AST. No "holes" are permitted. Of course, this is why I had you start with `Const` and `BinOp`, because those form a fully self-contained recursive structure in which each node can be assured to have `end_col_offset` set.

So it is actually more appropriate for you to have an `assert` statement enforcing this condition, rather than calling `set_end_col_offset_by_string`.